drivers/sensor/lsm6dso16is/lsm6dso16is_shub.c

```diff
@@ -0,0 +1,838 @@
+{
+	lsm6dso16is_status_master_t status;
+	int tries = 200; /* Should be max ~160 ms, from 2 cycles at slowest ODR 12.5 Hz */
+
+	do {
+		if (!--tries) {
+			LOG_DBG("shub: Timeout waiting for operation to complete");
+			return -ETIMEDOUT;
+		}
+		k_msleep(1);
+		lsm6dso16is_sh_status_mainpage_get(ctx, &status);
+	} while (status.sens_hub_endop == 0);
+
+	return 1;
+}
+
+static inline void lsm6dso16is_shub_embedded_en(stmdev_ctx_t *ctx, bool on)
+{
+	lsm6dso16is_func_cfg_access_t reg = {
+		.shub_reg_access = on ? LSM6DSO16IS_SENSOR_HUB_MEM_BANK :
+					LSM6DSO16IS_MAIN_MEM_BANK
+	};
+
+	lsm6dso16is_write_reg(ctx, LSM6DSO16IS_FUNC_CFG_ACCESS, (uint8_t *)&reg, 1);
+}
+
+static int lsm6dso16is_shub_read_embedded_regs(stmdev_ctx_t *ctx,
+					   uint8_t reg_addr,
+					   uint8_t *value, int len)
+{
+	lsm6dso16is_shub_embedded_en(ctx, true);
+
+	if (lsm6dso16is_read_reg(ctx, reg_addr, value, len) < 0) {
+		LOG_DBG("shub: failed to read external reg: %02x", reg_addr);
+		lsm6dso16is_shub_embedded_en(ctx, false);
+		return -EIO;
+	}
+
+	lsm6dso16is_shub_embedded_en(ctx, false);
+
+	return 0;
+}
+
+static int lsm6dso16is_shub_write_embedded_regs(stmdev_ctx_t *ctx,
+					    uint8_t reg_addr,
+					    uint8_t *value, uint8_t len)
+{
+	lsm6dso16is_shub_embedded_en(ctx, true);
+
+	if (lsm6dso16is_write_reg(ctx, reg_addr, value, len) < 0) {
+		LOG_DBG("shub: failed to write external reg: %02x", reg_addr);
+		lsm6dso16is_shub_embedded_en(ctx, false);
+		return -EIO;
+	}
+
+	lsm6dso16is_shub_embedded_en(ctx, false);
+
+	return 0;
+}
+
+static void lsm6dso16is_shub_enable(const struct device *dev, uint8_t enable)
+{
+	const struct lsm6dso16is_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm6dso16is_data *data = dev->data;
+
+	/* Enable Accel @26hz */
+	if (!data->accel_freq) {
+		uint8_t odr = (enable) ? 2 : 0;
+
+		if (lsm6dso16is_xl_data_rate_set(ctx, odr) < 0) {
+			LOG_DBG("shub: failed to set XL sampling rate");
+			return;
+		}
+	}
+
+	if (enable) {
+		lsm6dso16is_status_master_t status;
+
+		/* Clear any pending status flags */
+		lsm6dso16is_sh_status_mainpage_get(ctx, &status);
+	}
+
+	if (lsm6dso16is_sh_master_set(ctx, enable) < 0) {
+		LOG_DBG("shub: failed to set master on");
+		lsm6dso16is_shub_embedded_en(ctx, false);
+		return;
+	}
+
+	if (!enable) {
+		/* Necessary per AN5192 §7.2.1 */
+		k_busy_wait(300);
+	}
+}
+
+/* must be called with master on */
+static int lsm6dso16is_shub_check_slv0_nack(stmdev_ctx_t *ctx)
+{
+	uint8_t status;
+
+	if (lsm6dso16is_shub_read_embedded_regs(ctx, LSM6DSO16IS_SHUB_STATUS_MASTER,
+					     &status, 1) < 0) {
+		LOG_DBG("shub: error reading embedded reg");
+		return -EIO;
+	}
+
+	if (status & (LSM6DSO16IS_SHUB_STATUS_SLV0_NACK)) {
+		LOG_DBG("shub: SLV0 nacked");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/*
+ * use SLV0 for generic read to slave device
+ */
+static int lsm6dso16is_shub_read_slave_reg(const struct device *dev,
```

Comment:
Use an alternative to the word "slave"
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language